test/clj_puppetdb/core_test.clj

```diff
@@ -26,4 +28,64 @@
       ;; I'm testing for truthiness of conn here. Schema validation should handle the rest except the VCR piece,
       ;; and testing equality with java.io.File objects doesn't seem to work.
       (is conn)
-      (is (= "/temp" (get (conn) :vcr-dir))))))
+      (is (= (:vcr-dir opts) (:vcr-dir (conn))))
+      (is (:ssl-context (conn)))))
+  (testing "SSL connection should require certificates"
+    (is (thrown? IllegalArgumentException (connect "https://puppetdb:8081" {})))
+    (is (thrown? IllegalArgumentException (connect "https://puppetdb:8081"
+                                                   {:ssl-context "dummy"})))
+    (is (thrown? IllegalArgumentException (connect "https://puppetdb:8081"
+                                                   {:ssl-cert "dummy" :ssl-key "dummy" :ssl-ca-cert "dummy"}))))
+  (testing "Should do proper GET request"
+    (let [host "http://localhost:8080"
+          path "/v4/nodes"
+          params {:query [:= [:fact "operatingsystem"] "Linux"]}
+          url-params "?query=%5B%22%3D%22%2C%5B%22fact%22%2C%22operatingsystem%22%5D%2C%22Linux%22%5D"
+          conn (connect host)]
+      (with-redefs [http/get (fn[query _] (is (= query (str host path url-params))) [])]
+        (conn conn path params)
+        (conn (str host path url-params))))))
+
+(deftest query-test
+  (let [data ["node-01" "node-02" "node-03" "node-04"]
+        metadata {:total (count data)}
+        client "a client"
+        path "/v4/nodes"]
+    (testing "Should return data when parameters are not used"
+      (with-redefs [GET (fn[_ _ _]
+                          [data {"x-records" (.toString (count data))}])]
+        (is (= (query client path)
+               data))
+        (is (= (query-with-metadata client path {})
+               [data metadata]))))
+    (testing "Should return data when parameters are used"
+      (let [query-vec [:= [:fact "operatingsystem"] "Linux"]]
+        (with-redefs [GET (fn[_ _ params]
+                            (is (= (:query params) query-vec))
+                            [data {"x-records" (.toString (count data))}])]
+          (is (= (query client path query-vec)
+                 data))
+          (is (= (query-with-metadata client path query-vec {})
+                 [data metadata]))
+          (is (= (query-with-metadata client path {:query query-vec})
+                 [data metadata])))))))
+
+(deftest lazy-query-test
+  (testing "Should automatically fetch three pages by two items"
+    (let [data [["A" "B"] ["C" "D"] ["E" "F"]]]
+      (with-redefs [clj-puppetdb.http/GET (fn[_ _ {:keys [offset]}]
+                                            (let [idx (/ offset 2)]
+                                              (when (< idx 3) (nth data idx))))]
+        (let [seq (lazy-query nil nil {:limit 2 :order-by nil})]
+          (is (= (first seq) "A"))
+          (is (= (nth seq 4) "E"))
+          (is (= seq (flatten data)))))))
```

Comment:
The first two assertions must be satisfied if the third one is, i.e. the first two are superfluous.